include/unitTestMain.h

```diff
@@ -51,7 +89,38 @@ main(int, char *[])
     CPPUNIT_NS::CompilerOutputter outputter( &result, std::cerr );
     outputter.write();
 
-    return result.wasSuccessful() ? 0 : 1;
+    return result.wasSuccessful();
+}
+
+/// TestProgram object registered by RegisterTestProgram() (or nil).
+static std::unique_ptr<TestProgram> TestProgram_;
+
+/// Creates and registers a CustomTestProgram object. Optionally "uses" the
+/// static defined by RegisterTestProgram (by taking that variable address).
+/// Use RegisterTestProgram_() instead.
+template <class CustomTestProgram>
+static bool
+RegisterTestProgram_(void * = nullptr)
+{
+    // Avoid assert(): Some tests do not link with libcompatsquid.la (XXX?).
+    // assert(!TestProgram_);
+    if (TestProgram_)
+        throw std::runtime_error("attempt to register more than one TestProgram");
+
+    TestProgram_.reset(new CustomTestProgram());
+    return true;
+}
+
+/// A helper macro to create and register a CustomTestProgram object.
+#define RegisterTestProgram(CustomTestProgram) \
+    static bool CustomTestProgramRegistration_ = RegisterTestProgram_<CustomTestProgram>(&CustomTestProgramRegistration_)
```

Comment:
Note that these static initialization tricks are not going to be removed by LTO because they are located in the same translation unit as main() -- our test programs `#include` this file, and this file defines main(). Said that, if you know of a better way to handle this registration, please let me know. I could not find anything that would not require changing all test files, even those that do not need to customize startup/shutdown. If you think we should modify all of them to avoid these tricks, please let me know.

To see two (for now) usage examples of this macro outside of main(), grep this PR for RegisterTestProgram calls.

Reply:
> If you think we should modify all of them to avoid these tricks, please let me know.

I decided to remove these tricks, migrating away from "include main()" unit test design in commit 8908d8b. This comment details my reasoning:

As you know, official code defines main() in include/unitTestMain.h which is included by nearly all tests. This approach avoids duplicating main() definition (containing a simple single-line body) in each test file. However, placing main() in a header file:

* confuses new developers that are looking for main() as a well-known test program starting point; the code looks surprising or "magical";

* forces us to add clunky/risky/obscure RegisterTestProgram() interface (the subject of this comment thread) that uses static initialization side effects to "inject" additional functionality into TestProgram;

* violates basic developer expectations and Squid code style guidelines;

* complicates tests that need to customize their startup (about 25% of tests already contain some customization and, bugs notwithstanding, the portion of such tests is expected to _grow_). Arguably, these tests do not actually duplicate main() implementation because their main()s create a test-specific object rather than the "default" TestProgram object!

Tiny main() code duplication is the lesser evil here.